### PR TITLE
Add Code Sentiment post

### PR DESCRIPTION
**Pull request by vmarkovtsev**

### CODE REVIEW

**Comment by campoy** [comment]
Hey @vmarkovtsev, could you deploy this to staging?
It helps reviewing the looks too

**Comment by vmarkovtsev** [comment]
Sure. @dpordomingo I have never used staging for the blog, could you please point me to the instructions.

**Comment by dpordomingo** [comment]
Sure @vmarkovtsev Here is the guide https://github.com/src-d/guide/blob/master/engineering/continuous-delivery.md#deploying-into-staging

tl;dr
Just push your changes into `srcd/staging` branch.
Whenever that branch is pushed, it is deployed into staging env ;)

**Comment by vmarkovtsev** [comment]
@campoy Done https://blog-staging.srcd.run/post/codesent/

File: content/post/codesent.md [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+Structural clusters in Jinja2. Each dot is a function and each pair is closer proportionally to
+the number of times those two functions appeared in the same commit.
+{{% /caption %}}
+
+Under the cover, Hercules contains the DAG resolution engine which frees users from manually
+designing the analysis pipeline. It also leverages `pkg/plugin` to allow external extensions.
+
+We pass over to comment extraction. The most straightforward way to extract comments is to
+take a syntax highlighter with a bunch of regular expression rules. We will get into trouble once
+we want to distinguish docstrings from regular comments, so we need to parse the code fairly.
```
Comment (campoy) [inline content/post/codesent.md]: Not sure whether the word fairly is adequate in this context.
Maybe "correctly"?

File: content/post/codesent.md [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+in parallel for training and Go API and plain CPU for inference. It is impossible to train Tensorflow
+models using the current Go bindings, but even if it was possible, I would still not do that. From my
+experience, scripting languages are superior for doing research where you rewrite code every now and then
+and need the full expressiveness and flexibility to achieve the result as soon as possible.
+At the same time Go fits the inference scenario well: we put everything into the single executable
+without dependencies and it just works. Everybody who deployed models in Python should understand
+the pain.
+
+Have you heard about [neugram](https://neugram.io/), a scripting language on top of Go runtime?
+It was created by David Crawshaw, the author of `pkg/plugin`. Neugram is still in it's early ages
```
Comment (campoy) [inline content/post/codesent.md]: s/it's/its/

File: content/post/codesent.md [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+and has childhood problems, but I really like it. Who knows, maybe one day we will train Tensorflow
```
Comment (campoy) [inline content/post/codesent.md]: s/childhood problems/youth related issues/ ?

File: content/post/codesent.md [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+
+{{% caption src="/post/codesent/so.png" %}}
+StackOverflow question about the best comments.
+{{% /caption %}}
+
+I keep laughing every time I read the answers, it's a real pity this question is closed. So I thought
+that I could use the open source projects we've developed at source{d} to mine such comments in
+open source repositories. Most of the listed examples are sarcastic and belong to dark humour,
+and I decided to employ [sentiment analysis](https://en.wikipedia.org/wiki/Sentiment_analysis)
+as the NLP domain which is relatively well-studied and there exist big datasets for free download.
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply content/post/codesent.md]: So much better!

File: content/post/codesent.md [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+
+I will describe each one in detail now. Let's start with go-git.
+
+{{% caption src="/post/codesent/go-git.png" %}}
+go-git logo by Ricardo Baeta.
+{{% /caption %}}
+
+We started it back in 2015 as an "internal open source" project to fulfill our tasks of source code
+retrieval and analysis. This is an example of the typical software project workflow at source{d}
+today: we go open source from day 0 and see what happens next. This time go-git became a huge success
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply content/post/codesent.md]: argh, biased by 0-day

File: content/post/codesent.md [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+with more than 2400 stars and several companies already using it in production. Although it lacks
+some of the [features](https://github.com/src-d/go-git/blob/master/COMPATIBILITY.md) of "big brothers",
+it is performant and highly extensible. Finally, the maintainers are experienced code maniacs and
+go-git is a good example of idiomatic and clean Go.
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply content/post/codesent.md]: o-oh

[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply content/post/codesent.md]: From https://github.com/src-d/go-git/blob/master/CONTRIBUTING.md
```
All PRs must be written in idiomatic Go, formatted according to gofmt, and without any warnings from go lint nor go vet.
```
This was my foundation for this phrase. OK, will rephrase. After all, I love the API, whether it is idiomatic or not

File: content/post/codesent.md (hunk 3 above)
Comment (vmarkovtsev) [reply content/post/codesent.md]: Need to remember this phrase instead of copying shit from Russian, thanks!

**Comment by vmarkovtsev** [comment]
@campoy Thanks for the detailed and useful review, as always! Done.

File: content/post/codesent.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (mcuadros) [reply content/post/codesent.md]: Take as an example of idiomatic code `go-git` is saying too much, but of course, the code is idiomatic. I have written thousands of lines for dozens of projects, here and for other open source projects, I really believe is idiomatic. But not an example of nothing, for sure has a lot of parts that can be improved.  

Anyway, I wish to know which real project can be considered an example of this.

File: content/post/codesent.md [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+Structural clusters in Jinja2. Each dot is a function and each pair is closer proportionally to
+the number of times those two functions appeared in the same commit.
+{{% /caption %}}
+
+Under the cover, Hercules contains the DAG resolution engine which frees users from manually
+designing the analysis pipeline. It also leverages `pkg/plugin` to allow external extensions.
+
+### Babelfish
+
+We pass over to comment extraction. The most straightforward way to extract comments is to
```
Comment (eiso) [inline content/post/codesent.md]: I would avoid using "pass over" (https://www.google.com/search?q=pass+over+meaning&oq=%22pass+over%22+&aqs=chrome.1.69i57j0l5.2102j0j1&sourceid=chrome&ie=UTF-8). And say "We continue with comment extraction."

File: content/post/codesent.md [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+)
+client, _ := bblfsh.NewClient("0.0.0.0:9432")
+resp, _ := client.NewParseRequest().Content("...").Do()
+nodes, err := tools.Filter(resp.UAST, "//*[@roleComment]")
+// nodes[0].Token
+```
+
+As you see, barely 3 lines of code, error handling excluded, allow you to extract all the
+comments from a source code string. The cool thing is UAST pretending to be XML and running XPath
+quieries on it.
```
Comment (eiso) [inline content/post/codesent.md]: s/quieries/queries

File: content/post/codesent.md [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+graph execution from the actual hardware where it executes. Tensors may be big and operations on them
+benefit from various accelerators such as GPUs, and Tensorflow allows to pin each operation to the
+supported execution device.
+
+Each ML model goes through the training phase, either supervised or not. The graph state tensors change
+while the model is trained. Here is the intended way to train and use a typical neural network:
+
+1. Debug the model locally: make sure it converges and there are no implementation faults. We debug on [e-GPUs](https://egpu.io/) at source{d}.
+2. Train the model on as much data as you have on a GPU cluster.
+3. Optionally apply a metaparameter optimization algorithm to reach the best metrics in (2).
```
Comment (eiso) [inline content/post/codesent.md]: Give a link to a webpage that explains metaparameter optimization

File: content/post/codesent.md [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+4. Export the trained graph in "GraphDef" format (Protocol Buffers).
+5. Distribute it in an embedded or a pluggable fashion - that is, inside the application binary or separately.
+6. Apply it ("infer") with any programming language.
+
+In our case, I used Python API to Tensorflow on a dedicated machine with 4 GPUs running two experiments
```
Comment (eiso) [inline content/post/codesent.md]: s/I used Python API to Tensorflow /I used the Python API for Tensorflow/

File: content/post/codesent.md [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+in parallel for training and Go API and plain CPU for inference. It is impossible to train Tensorflow
```
Comment (eiso) [inline content/post/codesent.md]: s/and Go API/and the Go API/

File: content/post/codesent.md [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+and manage the environment. Technically, they are bridges to the underlying Tensorflow runtime
+written in C++ as a standalone library, in case with Go this bridge is implemented with CGo.
+
+### BiDiSentiment
+
+{{% caption src="/post/codesent/gophers.png" %}}
+vmarkovtsev/BiDiSentiment logo.
+{{% /caption %}}
+
+It's left to reveal the comment sentiment classification model. I named it "BiDiSentiment" and
```
Comment (eiso) [inline content/post/codesent.md]: s/It's left/What is left is/

File: content/post/codesent.md [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+The biggest labelled dataset that I found was 1.5 million tweets so I trained on it. As a bonus,
+the text sizes are bound to 140 (now 280) characters which perfectly suits the maximum recurrent memory
+depth. Here is the plot of the accuracy (number of correctly predicted sentiments divided by the overall
+size) through time.
+
+{{% caption src="/post/codesent/validation_black.png" %}}
+BiDiSentiment accuracy on validation.
+{{% /caption %}}
+
+We split the data into two parts: the part on which we train and the part on which we evaluate aka
```
Comment (eiso) [inline content/post/codesent.md]: s/aka/a.k.a/

File: content/post/codesent.md [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+validation. This is needed to avoid overfitting - when the model adapts to the training examples
+too much and loses generalization ability. The validation accuracy is plotted here. We see that
```
Comment (eiso) [inline content/post/codesent.md]: s/and loses/it loses/

File: content/post/codesent.md [hunk 15]
```diff
@@ -0,0 +1,395 @@
...
+our model overfits after 5 epochs - that is, runs across the whole training set. This is a typical
+situation in recurrent neural networks. I should have used overfit reduction techniques,
```
Comment (eiso) [inline content/post/codesent.md]: s/in recurrent/in a recurrent/

File: content/post/codesent.md [hunk 16]
```diff
@@ -0,0 +1,395 @@
...
+
+// TODO(bradfitz): this might be 0, once escape analysis is better
+```
+
+Source: [1](https://github.com/golang/go/blob/master/src/bytes/buffer.go#L63),
+[2](https://github.com/golang/go/blob/master/src/math/bits/bits.go#L135),
+[3](https://github.com/golang/go/blob/46e392e01c630dee41a67e01223b538aae9dc9b5/usr/austin/ogle/vars.go#L42),
+[4](https://github.com/golang/go/blob/84ef97b59c89b7d9fdc04a1a8a438cd3257bf521/src/pkg/strconv/strconv_test.go#L21).
+
+The last one is actually written by Brad who talked a few minutes ago hehe. In this case the model
```
Comment (eiso) [inline content/post/codesent.md]: s/who talked a few minutes ago hehe./who gave the Gophercon talk right before me hehe./

File: content/post/codesent.md [hunk 17]
```diff
@@ -0,0 +1,395 @@
...
+# With TensorFlow, we can infer the output shape directly
+
+# Make sure there is exactly one 1 in a row
+
+# get sorted list of layer depths
+
+# Theano has a built-in optimization for logsumexp so we can just write the expression directly
+```
+
+I have looked at other classification result and can conclude the following. First, the model
```
Comment (eiso) [inline content/post/codesent.md]: s/result/results/

File: content/post/codesent.md [hunk 18]
```diff
@@ -0,0 +1,395 @@
...
+was trained on data of one nature (tweets) and applied to data of different nature (comments).
```
Comment (eiso) [inline content/post/codesent.md]: s/of different/of a different/

File: content/post/codesent.md (hunk 7 above)
Comment (vmarkovtsev) [reply content/post/codesent.md]: LOL. I want to kill my teachers now

File: content/post/codesent.md (hunk 15 above)
Comment (vmarkovtsev) [reply content/post/codesent.md]: But it's networks :thinking:

File: content/post/codesent.md (hunk 16 above)
Comment (vmarkovtsev) [reply content/post/codesent.md]: It was Brad then Brian then me...

File: content/post/codesent.md (hunk 13 above)
Comment (vmarkovtsev) [reply content/post/codesent.md]: Wiki and Cambridge say aka is OK https://en.wikipedia.org/wiki/Aka https://dictionary.cambridge.org/dictionary/english/aka

**Comment by vmarkovtsev** [comment]
@eiso Thanks, I have applied the fixes

File: content/post/codesent.md [hunk 19]
```diff
@@ -0,0 +1,396 @@
...
+written in C++ as a standalone library, in case with Go this bridge is implemented with CGo.
+
+### BiDiSentiment
+
+{{% caption src="/post/codesent/gophers.png" %}}
+vmarkovtsev/BiDiSentiment logo.
+{{% /caption %}}
+
+What's left is to reveal the comment sentiment classification model. I named it "BiDiSentiment" and
+this is it's logo. BiDiSentiment is a general-purpose sentiment model, it is not specifically designed
```
Comment (campoy) [inline content/post/codesent.md]: s/it's/its/

Comment (campoy) [inline content/post/codesent.md]: Why `BiDiSentiment`? Is there something related to Bidirectional? What does it mean?

Comment (campoy) [reply content/post/codesent.md]: Oh, this is bidirectional recurrent network? use the term explicitly

File: content/post/codesent.md [hunk 20]
```diff
@@ -0,0 +1,396 @@
...
+The biggest labelled dataset that I found was 1.5 million tweets so I trained on it. As a bonus,
+the text sizes are bound to 140 (now 280) characters which perfectly suits the maximum recurrent memory
+depth. Here is the plot of the accuracy (number of correctly predicted sentiments divided by the overall
+size) through time.
+
+{{% caption src="/post/codesent/validation_black.png" %}}
+BiDiSentiment accuracy on validation.
+{{% /caption %}}
+
+We split the data into two parts: the part on which we train and the part on which we evaluate aka
```
Comment (campoy) [inline content/post/codesent.md]: I think you should use the well established terms here: training and validation datasets.

We split our full dataset into two parts:

- the training dataset contains X% of randomly selected samples, and it's used to train our model,
- the validation dataset contains the remaining %X and it's used to evaluate the model's accuracy.

File: content/post/codesent.md [hunk 21]
```diff
@@ -0,0 +1,396 @@
...
+validation. This is needed to avoid overfitting - when the model adapts to the training examples
+too much it loses generalization ability. The validation accuracy is plotted here. We see that
+our model overfits after 5 epochs - that is, runs across the whole training set. This is a typical
```
Comment (campoy) [inline content/post/codesent.md]: after 5 epochs. An epoch is completed after the model has seen the whole training dataset once.

File: content/post/codesent.md [hunk 22]
```diff
@@ -0,0 +1,396 @@
...
+Let's see how BiDiSentiment performs on some answers to that StackOverflow questions.
+
+```
+go get -v gopkg.in/vmarkovtsev/BiDiSentiment.v1/...
+
+echo "When I wrote this, only God and I understood \
+what I was doing. Now, God only knows" | $GOPATH/bin/sentiment
+```
+
+Output (negative probability): 0.8803515 - the sentiment is clearly negative.
```
Comment (campoy) [inline content/post/codesent.md]: Could you also give how the sample was originally labeled?

File: content/post/codesent.md [hunk 23]
```diff
@@ -0,0 +1,396 @@
...
+| [pygame/pygame](https://github.com/pygame/pygame)         | 93.4     | 98.2     | -4.8      |
+| [pallets/flask](https://github.com/pallets/flask)         | 23.6     | 31.9     | -8.3      |
+| [django/django](https://github.com/django/django)         | 462.1    | 536.1    | -74.0     |
+| [golang/go](https://github.com/golang/go)             | 402.1    | 507.3    | -105.2    |
+| [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes) | 186.1    | 91.6     | 94.5      |
+| [keras-team/keras](https://github.com/keras-team/keras)      | 104.8    | 71.1     | 33.7      |
+
+The columns represent the sum of all the positive and negative sentiments together with the final
+sentiment value. Some of the projects appeared to be negative and some positive. We'll see why
+later, and now let's look at the promised sentiment plot through time.
```
Comment (campoy) [inline content/post/codesent.md]: > later.
> 
> Let's look at the promised sentiment plot through time now!

File: content/post/codesent.md [hunk 24]
```diff
@@ -0,0 +1,396 @@
...
+
+{{% caption src="/post/codesent/golang-sentiment-black.png" %}}
+golang/go sentiment through time.
+{{% /caption %}}
+
+Can we conclude anything useful from this? Unfortunately, no, there are no clear trends or
+periods. It is too many different comments. Maybe we will have more luck with monorepos, that is
```
Comment (campoy) [inline content/post/codesent.md]: There are too many different comments? Not sure what you mean here

File: content/post/codesent.md [hunk 25]
```diff
@@ -0,0 +1,396 @@
...
+those with a single main contributor. For now, it is more fun to inspect the most negative and positive
```
Comment (campoy) [inline content/post/codesent.md]: that's not what a monorepo is, a monorepo is a repository that has all of the code for a company, such as Google

avoid the term here, instead call it "single contributor repo" or something like that

File: content/post/codesent.md [hunk 26]
```diff
@@ -0,0 +1,396 @@
...
+examples of comments on Golang source code. Negative:
```
Comment (campoy) [inline content/post/codesent.md]: s/Golang/Go/

File: content/post/codesent.md [hunk 27]
```diff
@@ -0,0 +1,396 @@
...
+```
+
+I have looked at other classification results and can conclude the following. First, the model
+was trained on data of one nature (tweets) and applied to data of a different nature (comments).
+Thus it often exaggerates comment sentiment. Second, it behaves differently on two project types.
+The first type is projects which contain many details, many conditions, many checks and
+handle many edge cases. For example, language compilers (Go) or web servers (Django). BiDiSentiment
+classifies them as negative: the text tells how **not** to fail.
+The second type is projects which explain what is happening, which environment features they use
+to do better or explain much domain terminology. For example, k8s has tons of devops comments
```
Comment (campoy) [inline content/post/codesent.md]: s/k8s/Kubernetes/g

File: content/post/codesent.md [hunk 28]
```diff
@@ -0,0 +1,396 @@
...
+and Keras describes deep learning recipes. BiDiSentiment treats them as positive: they teach or
+explain why they are cool. I must say that deep learning seems to be the modern alchemy
+([discussion](https://medium.com/@Synced/lecun-vs-rahimi-has-machine-learning-become-alchemy-21cb1557920d))
+so Keras has a big positive sentiment, hehe.
+
+## Conclusion
+
+* It is easy to analyse Git repositories with Go.
```
Comment (campoy) [inline content/post/codesent.md]: s/analyse/analyze/ (US spelling)

File: content/post/codesent.md [hunk 29]
```diff
@@ -0,0 +1,396 @@
...
+* Go is handy for inference-in-a-box with Tensorflow.
```
Comment (campoy) [inline content/post/codesent.md]: `inference-in-a-box`
is this a well known term?
Maybe something like: Go provides an easy deployment of tools performing inference on Tensorflow models.

File: content/post/codesent.md [hunk 30]
```diff
@@ -0,0 +1,396 @@
...
+* We need a dedicated dataset for comment sentiment. At the same time, labelling 1 million comments
+is at least challenging.
```
Comment (campoy) [inline content/post/codesent.md]: IDEA: could we use the github reactions on comments as a way to pseudo label comments?

File: content/post/codesent.md [hunk 31]
```diff
@@ -0,0 +1,396 @@
...
+* Naive sentiment alignment depends on the project type.
```
Comment (campoy) [inline content/post/codesent.md]: I don't understand this line. sentiment alignment?
Many people will read *only* the conclusions, so make sure they're so clear and impressive that people will want to read the rest of the article

File: content/post/codesent.md (hunk 19 above)
Comment (vmarkovtsev) [reply content/post/codesent.md]: This is the name of my project. Now it's evident that I should link it to vmarkovtsev/bidisentiment :) It was in the presentation.

File: content/post/codesent.md (hunk 22 above)
Comment (vmarkovtsev) [reply content/post/codesent.md]: This is https://stackoverflow.com/a/316233 there is not label :)

File: content/post/codesent.md (hunk 25 above)
Comment (vmarkovtsev) [reply content/post/codesent.md]: I used this a few times as "single contrib repo" before. As you've already discovered I like giving short names to frequent terms, and "monorepo" was one of them back in 2016-2017. No objections to changing it though.

File: content/post/codesent.md (hunk 29 above)
Comment (vmarkovtsev) [reply content/post/codesent.md]: This will be the name I will do my best to push to the community

File: content/post/codesent.md (hunk 30 above)
Comment (vmarkovtsev) [reply content/post/codesent.md]: Do you mean issues? Nice one!

File: content/post/codesent.md (hunk 31 above)
Comment (vmarkovtsev) [reply content/post/codesent.md]: Yeah, I took this from RPG games. "alignment" there is either good (elves, humans, etc.) or evil (necro, inferno, dark elves). Let's rephrase

File: content/post/codesent.md (hunk 30 above)
Comment (eiso) [reply content/post/codesent.md]: @campoy we're in sync lately: https://src-d.slack.com/archives/C7USX021L/p1521810248000012?thread_ts=1521805187.000215&cid=C7USX021L

**Comment by vmarkovtsev** [comment]
@campoy Done

**Pull request merged**